### PR TITLE
optimized some ignore rules in  the `phpstan-baseline.neon`

### PR DESCRIPTION
**Pull request by TheNorthMemory**
- Improved the codes and comments, cleanup the ignore rules in the `phpstan-baseline.neon`;
- The `pcre.backtrack_limit` settings may be affected  the function `preg_split`'s behavior. While there were a customized value (less than 5), it can caused the `Header::parse` and `Header::normalize` working incorrectly. Optimize codes for this case.

### CODE REVIEW

File: tests/HeaderTest.php [hunk 1]
```diff
@@ -68,4 +68,13 @@ public function testParsesArrayHeaders(): void
         $header = ['a, b', 'c', 'd, e'];
         self::assertSame(['a', 'b', 'c', 'd', 'e'], Psr7\Header::normalize($header));
     }
+
+    public function testNormalizeWithPcreBackTrackLimited(): void
+    {
+        $limited = ini_get('pcre.backtrack_limit');
+        ini_set('pcre.backtrack_limit', '4');
+        $things = ['"Google Chrome";v="95", "Chromium";v="95", ";Not A Brand";v="99"'];
+        self::assertSame(['"Google Chrome";v="95", "Chromium";v="95", ";Not A Brand";v="99"'], Psr7\Header::normalize($things));
```
Comment (TimWolla) [inline tests/HeaderTest.php:77]: This assertion is bogus. However the backtrack limit is a good reason to not rely on regular expressions here. See: #447

Comment (TheNorthMemory) [reply tests/HeaderTest.php:77]: Agree. Rely on the regular expression may not a good idea. For example, parsing the new [Permissions-Policy](https://github.com/w3c/webappsec-permissions-policy/blob/main/permissions-policy-explainer.md#new-header)'s value, eg: `Permissions-Policy: fullscreen=(self "https://example.com" "https://another.example.com"), geolocation=*, camera=()`, there were the parentheses, `self`, `none`, empty string and double quotes strings, it shall be so hard to do that.

File: src/Uri.php [hunk 2]
```diff
@@ -585,11 +585,14 @@ private function filterUserInfoComponent($component): string
             throw new \InvalidArgumentException('User info must be a string');
         }
 
-        return preg_replace_callback(
+        /** @var string $thing */
```
Comment (GrahamCampbell) [inline src/Uri.php:588]: No need to do an assignment here. Just add `/** @var string */` above the return statement.

File: src/Uri.php [hunk 3]
```diff
@@ -681,11 +684,14 @@ private function filterPath($path): string
             throw new \InvalidArgumentException('Path must be a string');
         }
 
-        return preg_replace_callback(
+        /** @var string $thing */
```
Comment (GrahamCampbell) [inline src/Uri.php:687]: same

File: src/Uri.php [hunk 4]
```diff
@@ -701,11 +707,14 @@ private function filterQueryAndFragment($str): string
             throw new \InvalidArgumentException('Query and fragment must be a string');
         }
 
-        return preg_replace_callback(
+        /** @var string $thing */
```
Comment (GrahamCampbell) [inline src/Uri.php:710]: same

File: src/Utils.php [hunk 5]
```diff
@@ -14,14 +14,14 @@ final class Utils
     /**
      * Remove the items given by the keys, case insensitively from the data.
      *
-     * @param string[] $keys
+     * @param (string|int)[] $keys
```
Comment (GrahamCampbell) [inline src/Utils.php:17]: please revert all changes to this. we know this is an issue, and the phpdoc intentionally does not allow ints.

File: src/StreamWrapper.php [hunk 6]
```diff
@@ -44,7 +44,10 @@ public static function getResource(StreamInterface $stream)
                 . 'writable, or both.');
         }
 
-        return fopen('guzzle://stream', $mode, false, self::createStreamContext($stream));
+        /** @var resource $handle */
```
Comment (GrahamCampbell) [inline src/StreamWrapper.php:47]: same

File: src/UploadedFile.php [hunk 7]
```diff
@@ -113,6 +113,9 @@ private function setError(int $error): void
         $this->error = $error;
     }
 
+    /**
+     * @param string|mixed $param
```
Comment (GrahamCampbell) [inline src/UploadedFile.php:117]: There are a few places where you made this mistake. Never include mixed in a union type. It already includes every other type (apart from void and never). `string|mixed` is just `mixed`.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Utils.php (hunk 5 above)
Comment (TheNorthMemory) [reply src/Utils.php:17]: > 

Oh, I see. It looks like designing for string `keys` removal only. Let me revert it.

File: src/StreamWrapper.php [hunk 8]
```diff
@@ -44,9 +44,9 @@ public static function getResource(StreamInterface $stream)
                 . 'writable, or both.');
         }
 
-        /** @var resource $handle */
         $handle = fopen('guzzle://stream', $mode, false, self::createStreamContext($stream));
```
Comment (GrahamCampbell) [inline src/StreamWrapper.php:47]: Please remove the extra assignment you added

File: src/Uri.php [hunk 9]
```diff
@@ -585,13 +585,13 @@ private function filterUserInfoComponent($component): string
             throw new \InvalidArgumentException('User info must be a string');
         }
 
-        /** @var string $thing */
         $thing = preg_replace_callback(
```
Comment (GrahamCampbell) [inline src/Uri.php:588]: Please remove the extra assignment you added

File: src/Uri.php [hunk 10]
```diff
@@ -684,13 +684,13 @@ private function filterPath($path): string
             throw new \InvalidArgumentException('Path must be a string');
         }
 
-        /** @var string $thing */
         $thing = preg_replace_callback(
```
Comment (GrahamCampbell) [inline src/Uri.php:687]: etc, in all places

File: src/UriNormalizer.php [hunk 11]
```diff
@@ -207,11 +209,13 @@ private static function decodeUnreservedCharacters(UriInterface $uri): UriInterf
...
         };
 
-        /** @var string $pathStr */
         $pathStr = preg_replace_callback($regex, $callback, $uri->getPath());
-        /** @var string $queryStr */
         $queryStr = preg_replace_callback($regex, $callback, $uri->getQuery());
 
+        /**
+         * @var string $pathStr
+         * @var string $queryStr
```
Comment (GrahamCampbell) [inline src/UriNormalizer.php:217]: What you had before was fine in this instance.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/UriNormalizer.php (hunk 11 above)
Comment (TheNorthMemory) [reply src/UriNormalizer.php:217]: Ok, let me try my best.

File: src/Uri.php (hunk 10 above)
Comment (GrahamCampbell) [reply src/Uri.php:687]: no, it means adding the var over the return statement itself. you dont need to name the variable.

Comment (TheNorthMemory) [reply src/Uri.php]: I see. Please take a review again.

**Comment by GrahamCampbell** [comment]
There are a lot of changes here, and it's not clear if any bugs are actually fixed. Since there are merge conflicts, we're not able to merge at this time. Please feel free to re-submit a PR fixing a specific bug.

**Pull request closed without merging**